modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+            }
+            if (withEdgeNGrams) {
+                tokenStream = new EdgeNGramTokenFilter(tokenStream, Defaults.MIN_GRAM, Defaults.MAX_GRAM, true);
+            }
+            return new TokenStreamComponents(components.getTokenizer(), tokenStream);
+        }
+
+        public boolean isWithEdgeNGrams() {
+            return withEdgeNGrams;
+        }
+
+        public int getShingleSize() {
+            return shingleSize;
+        }
+
+        public boolean isWithShingles() {
+            return withShingles;
+        }
+    }
+
+    public static class SuggesterizedFieldType extends TermBasedFieldType {
+
+        SuggesterizedFieldType(String name) {
+            setName(name);
+            setOmitNorms(true);
+            setIndexOptions(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS); // todo do we need to always set omitNorms and INdexOptions
+            setTokenized(true);
+        }
+
+        SuggesterizedFieldType(SuggesterizedFieldType reference) {
+            super(reference);
+        }
+
+        @Override
+        public SuggesterizedFieldType clone() {
+            return new SuggesterizedFieldType(this);
+        }
+
+        @Override
+        public String typeName() {
+            return SuggesterizedFieldMapper.CONTENT_TYPE;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
+        }
+    }
+
+    public static class SuggesterizedFieldMapper extends FieldMapper implements ArrayValueMapperParser { // todo better name
+
+        static final String CONTENT_TYPE = "suggesterized";
+
+        protected SuggesterizedFieldMapper(SuggesterizedFieldType fieldType, Settings indexSettings) {
+            super(fieldType.name(), fieldType, fieldType, indexSettings, MultiFields.empty(), CopyTo.empty());
+        }
+
+        @Override
+        public SuggesterizedFieldType fieldType() {
+            return (SuggesterizedFieldType) super.fieldType();
+        }
+
+        void addField(String value, List<IndexableField> fields) {
+            fields.add(new Field(fieldType().name(), value, fieldType));
+        }
+
+        @Override
+        protected void parseCreateField(ParseContext context, List<IndexableField> fields) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected String contentType() {
+            return CONTENT_TYPE;
+        }
+    }
+
+    public static class SearchAsYouTypeFieldType extends TermBasedFieldType {
+
+        private Set<SuggesterizedFieldType> suggesterizedFieldTypes; // todo we might need a more structured way to know which
+        // field type is which
+        // eg we could make a SuggesterizedFieldTypeGroup class that can look up the only edge ngreams field and then also do
+        // number of shingles, true/false edge ngrams
+        // the alternative is do a bunch of string math
```

Comment:
+1 for a simple way to retrieve the shingle and edge ngram fields easily. We should avoid string math ;).